File: skills/visibility-architect/scripts/pagespeed-audit.ts
#!/usr/bin/env bun

/**
 * PageSpeed Audit Utility (Bun)
 * Helps the Visibility Architect audit core web vitals and SEO performance.
 * 
 * Usage: PAGESPEED_API_KEY=<your_api_key> bun pagespeed-audit.ts <url> [strategy: mobile|desktop]
 * 
 * Note for AI Agents: If the API key is missing, ask the user to provide it or check the .env file.
 */

const apiKey = process.env.PAGESPEED_API_KEY;

// Parse args: supports flags (--url X, --strategy=Y) and legacy positional args.
const USAGE = 'Usage: bun pagespeed-audit.ts [--url <url>] [--strategy mobile|desktop]\nAlso accepted positionally: bun pagespeed-audit.ts <url> [strategy]';

function parseArgs(argv: string[]): { url: string | null; strategy: string } {
  const opts: { url: string | null; strategy: string | null } = { url: null, strategy: null };
  const positional: string[] = [];
  for (let i = 0; i < argv.length; i++) {
    const arg = argv[i]!;
    if (arg === '--url') {
      opts.url = argv[++i] ?? null;
    } else if (arg.startsWith('--url=')) {
      opts.url = arg.slice('--url='.length);
    } else if (arg === '--strategy') {
      opts.strategy = argv[++i] ?? null;
    } else if (arg.startsWith('--strategy=')) {
      opts.strategy = arg.slice('--strategy='.length);
    } else if (arg === '-h' || arg === '--help') {
      console.log(USAGE);
      process.exit(0);
    } else if (arg.startsWith('-')) {
      console.error(`Unknown option: ${arg}\n${USAGE}`);
      process.exit(1);
    } else {
      positional.push(arg);
    }
  }
  if (!opts.url && positional.length > 0) opts.url = positional[0]!;
  if (!opts.strategy && positional.length > 1) opts.strategy = positional[1]!;
  const strategy = (opts.strategy || 'mobile').toLowerCase();
  if (!['mobile', 'desktop'].includes(strategy)) {
    console.error(`Invalid strategy: ${strategy} (must be 'mobile' or 'desktop')`);
    process.exit(1);
  }
  return { url: opts.url, strategy };
}

const { url, strategy } = parseArgs(process.argv.slice(2));

if (!url) {
  console.error(USAGE);
  console.error('Example: PAGESPEED_API_KEY=AIzaSy... bun pagespeed-audit.ts --url https://example.com --strategy desktop');
  process.exit(1);
}

if (!apiKey) {
  console.error('Error: PAGESPEED_API_KEY environment variable is missing.');
  console.error('Agent Action Required: Please ask the user to provide their Google PageSpeed API key, or read it from a .env file, then rerun the command with the key.');
  process.exit(1);
}

const apiUrl = `https://www.googleapis.com/pagespeedonline/v5/runPagespeed?url=${encodeURIComponent(url)}&strategy=${strategy.toUpperCase()}&category=performance&category=seo&category=accessibility&category=best-practices&key=${apiKey}`;

console.log(`Auditing ${url} (${strategy} strategy)...`);

try {
  const response = await fetch(apiUrl);
  if (!response.ok) {
    throw new Error(`HTTP error! status: ${response.status}`);
  }

  const result = await response.json();

  if (result.error) {
    console.error('Error from API:', result.error.message);
    process.exit(1);
  }

  const categories = result.lighthouseResult.categories;
  const audits = result.lighthouseResult.audits;
  const catScore = (id: string): number | null =>
    categories[id] && categories[id].score !== null ? Math.round(categories[id].score * 100) : null;

  const performanceScore = catScore('performance');
  const seoScore = catScore('seo');
  const a11yScore = catScore('accessibility');
  const bpScore = catScore('best-practices');

  console.log('\n--- PageSpeed Insights Report (Bun) ---');
  console.log(`URL: ${url}`);
  console.log(`Strategy: ${strategy}`);
  console.log(`Performance Score: ${performanceScore}  (target: 90+)`);
  console.log(`SEO Score: ${seoScore}  (target: 90+)`);
  if (a11yScore !== null) console.log(`Accessibility Score: ${a11yScore}  (target: 90+)`);
  if (bpScore !== null) console.log(`Best Practices Score: ${bpScore}  (target: 90+)`);

  // CrUX field data (real-user Core Web Vitals, 28-day window)
  const le = result.loadingExperience || {};
  const leMetrics = le.metrics || {};
  if (Object.keys(leMetrics).length > 0) {
    console.log(`\nCore Web Vitals (Field Data, ${le.overall_category || 'n/a'}):`);
    for (const [id, m] of Object.entries(leMetrics)) {
      console.log(`- ${id}: ${m.percentile} (${m.category})`);
    }
  } else {
    console.log('\nCore Web Vitals (Field Data): no CrUX data (traffic below threshold) — relying on lab data below');
  }

  console.log('\nKey Metrics (Lab):');
  console.log(`- First Contentful Paint: ${metricsVal(audits, 'first-contentful-paint')}`);
  console.log(`- Largest Contentful Paint: ${metricsVal(audits, 'largest-contentful-paint')}`);
  console.log(`- Cumulative Layout Shift: ${metricsVal(audits, 'cumulative-layout-shift')}`);
  console.log(`- Total Blocking Time: ${metricsVal(audits, 'total-blocking-time')}`);
  console.log(`- Speed Index: ${metricsVal(audits, 'speed-index')}`);

  // Failing audits (score < 0.9) across all categories, excluding pass-only/informational audits
  const EXCLUDE = new Set(['metrics', 'screenshot-thumbnails', 'final-screenshot', 'timing', 'network-requests', 'network-rtt', 'network-server-latency', 'diagnostics', 'debug-data', 'stacks', 'errors-in-console', 'worker-timing', 'lcp-lazy-loaded', 'prioritize-lcp-image', 'uses-rel-preconnect', 'third-party-summary', 'first-contentful-paint', 'largest-contentful-paint', 'cumulative-layout-shift', 'total-blocking-time', 'speed-index', 'interactive']);
  const failing: { id: string; score: number; displayValue: string; weighted: boolean }[] = [];
  for (const [id, audit] of Object.entries(audits) as [string, any][]) {
    if (audit.score === null || audit.score === undefined || audit.score >= 0.9 || EXCLUDE.has(id)) continue;
    if (audit.scoreMode !== 'binary' && audit.scoreMode !== 'metricSavings') continue;
    const weighted = Object.values(categories).some((c: any) => c.auditRefs && c.auditRefs.some((r: any) => r.id === id && r.weight > 0));
    failing.push({ id, score: audit.score, displayValue: audit.displayValue || '', weighted });
  }
  failing.sort((a, b) => a.score - b.score);
  if (failing.length > 0) {
    console.log('\nFailing Audits (score < 0.9):');
    for (const f of failing.slice(0, 15)) {
      const tag = f.weighted ? ' [core-weighted]' : '';
      console.log(`- ${f.id}: ${f.displayValue || 'failed'}${tag}`);
    }
  }

  // Savings opportunities (bytes/ms estimates)
  const SAVINGS = ['unused-javascript', 'unused-css-rules', 'render-blocking-resources', 'modern-image-formats', 'uses-optimized-images', 'offscreen-images', 'uses-responsive-images', 'uses-long-cache-ttl', 'legacy-javascript', 'duplicated-javascript', 'font-display'];
  const opportunities = SAVINGS
    .map(id => audits[id])
    .filter((a: any) => a && a.score !== null && a.score !== undefined && a.score < 1)
    .sort((a: any, b: any) => a.score - b.score);
  if (opportunities.length > 0) {
    console.log('\nSavings Opportunities:');
    for (const o of opportunities) {
      console.log(`- ${o.id}: ${o.displayValue || 'see report'}`);
    }
  }
  console.log('---------------------------------------\n');

  const scores = [performanceScore, seoScore, a11yScore, bpScore].filter((s): s is number => s !== null);
  if (scores.some(s => s < 90)) {
    const weak = scores.filter(s => s < 90).length;
    console.log(`Action Item: ${weak} categor${weak === 1 ? 'y is' : 'ies are'} below the 90 target. Advise the Experience Architect to prioritize the failing audits above.`);
  } else {
    console.log('Status: All category scores meet the 90+ target. Visibility targets met.');
  }

} catch (err) {
  console.error('Audit failed:', err instanceof Error ? err.message : err);
  process.exit(1);
}

function metricsVal(audits: Record<string, any>, id: string): string {
  return (audits[id] && audits[id].displayValue) || 'n/a';
}

File: skills/visibility-architect/scripts/pagespeed-audit.py
#!/usr/bin/env python3

"""
PageSpeed Audit Utility (Python)
Helps the Visibility Architect audit core web vitals and SEO performance.

Usage: PAGESPEED_API_KEY=<your_api_key> python pagespeed-audit.py <url> [strategy: mobile|desktop]

Note for AI Agents: If the API key is missing, ask the user to provide it or check the .env file.
"""

import sys
import os
import urllib.request
import urllib.parse
import json

# Remediation guide: audit id -> concise, actionable fix instruction.
FIXES = {
    # Accessibility
    "color-contrast": "Ensure text/background contrast >= 4.5:1 (3:1 for large text). Darken light-gray text (e.g. #a4a4a4 -> #767676) and avoid white text on light backgrounds. Verify with a WCAG contrast checker.",
    "link-name": "Add accessible names to icon/image-only links: aria-label=\"Facebook Nakhla Tour\" or visually-hidden text inside the <a>.",
    "heading-order": "Do not skip heading levels. Wrap page content in one H1, then H2 > H3 > H4 in order; use CSS classes for visual sizing instead of jumping tags.",
    "link-in-text-block": "Make inline links distinguishable: add underline, bold weight, or >= 3:1 color contrast versus surrounding text.",
    "image-aspect-ratio": "Set width/height attributes on <img> matching the intrinsic aspect ratio, or CSS aspect-ratio, so the browser reserves correct space.",
    "unsized-images": "Add explicit width and height attributes (or CSS aspect-ratio) to every <img> to prevent layout shifts.",
    "button-name": "Add aria-label or inner text to icon-only buttons.",
    "document-title": "Add a descriptive <title> to the page.",
    "html-has-lang": "Set lang attribute on <html> (e.g. lang=\"id\").",
    "tap-targets": "Make touch targets at least 48x48 CSS px with 8px spacing between them.",
    # Performance diagnostics
    "image-delivery-insight": "Serve images in WebP/AVIF, right-sized to display dimensions (srcset/sizes), compressed, and lazy-loaded below the fold. If behind Cloudflare, enable Polish + WebP.",
    "cache-insight": "Serve versioned static assets (/assets/*) with Cache-Control: public, max-age=31536000, immutable.",
    "unused-javascript": "Remove or defer unused JS: code-split bundles, tree-shake, or dynamically import non-critical modules (carousels, modals).",
    "unused-css-rules": "Purge unused CSS rules per page or remove blocking stylesheets not used above the fold.",
    "render-blocking-insight": "Inline critical CSS in <head> and defer the rest (media=\"print\" onload pattern), or reduce blocking stylesheet count.",
    "render-blocking-resources": "Inline critical CSS; defer non-critical stylesheets and scripts (defer/async).",
    "font-display-insight": "Add font-display: swap to @font-face declarations and preload primary font files.",
    "font-display": "Add font-display: swap (or optional) to @font-face rules.",
    "legacy-javascript-insight": "Raise the JS build target (drop ES5 polyfills/transforms for modern browsers), or serve legacy bundles only via nomodule.",
    "legacy-javascript": "Update build targets to ES2017+ and split legacy bundles behind nomodule.",
    "forced-reflow-insight": "Batch DOM reads before writes; avoid repeated style reads interleaved with style writes (layout thrashing).",
    "network-dependency-tree-insight": "Shorten the critical request chain: preconnect to third-party origins, self-host critical fonts/CSS, combine small files.",
    "third-party-summary": "Audit third-party scripts; defer non-essential ones and remove duplicates.",
    "modern-image-formats": "Convert JPEG/PNG to WebP or AVIF (<picture> element or CDN auto-negotiation).",
    "uses-optimized-images": "Compress images (lossy quality ~80) before upload or via CDN transforms.",
    "offscreen-images": "Add loading=\"lazy\" to below-the-fold images; never lazy-load the LCP/hero image.",
    "uses-responsive-images": "Provide multiple sizes via srcset + sizes so mobile devices do not download desktop-sized images.",
    "uses-long-cache-ttl": "Add long Cache-Control max-age with immutable to fingerprinted static assets.",
    "prioritize-lcp-image": "Preload the LCP image: <link rel=\"preload\" as=\"image\" fetchpriority=\"high\"> and avoid lazy-loading it.",
    "uses-rel-preconnect": "Add <link rel=\"preconnect\"> for critical third-party origins (CDN, font hosts).",
    "redirects": "Eliminate redirect chains; link directly to the final URL.",
    "total-byte-weight": "Reduce page weight: smaller images, fewer fonts, trim payloads.",
    "duplicated-javascript": "Dedupe shared modules into a common chunk instead of bundling them per-page.",
    # Best practices
    "errors-in-console": "Fix runtime errors reported in the browser console.",
    "no-vulnerable-libraries": "Upgrade JS dependencies with known CVEs.",
    "csp-xss": "Add a Content-Security-Policy header to mitigate XSS.",
}

def parse_args(argv):
    """Parse args: supports flags (--url X, --strategy=Y) and legacy positional args."""
    usage = (
        "Usage: python pagespeed-audit.py [--url <url>] [--strategy mobile|desktop]\n"
        "Also accepted positionally: python pagespeed-audit.py <url> [strategy]"
    )
    opts = {"url": None, "strategy": None}
    positional = []
    i = 0
    while i < len(argv):
        arg = argv[i]
        if arg == "--url":
            i += 1
            opts["url"] = argv[i] if i < len(argv) else None
        elif arg.startswith("--url="):
            opts["url"] = arg[len("--url="):]
        elif arg == "--strategy":
            i += 1
            opts["strategy"] = argv[i] if i < len(argv) else None
        elif arg.startswith("--strategy="):
            opts["strategy"] = arg[len("--strategy="):]
        elif arg in ("-h", "--help"):
            print(usage)
            sys.exit(0)
        elif arg.startswith("-"):
            print(f"Unknown option: {arg}\n{usage}", file=sys.stderr)
            sys.exit(1)
        else:
            positional.append(arg)
        i += 1
    if not opts["url"] and positional:
        opts["url"] = positional[0]
    if not opts["strategy"] and len(positional) > 1:
        opts["strategy"] = positional[1]
    strategy = (opts["strategy"] or "mobile").lower()
    if strategy not in ("mobile", "desktop"):
        print(f"Invalid strategy: {strategy} (must be 'mobile' or 'desktop')", file=sys.stderr)
        sys.exit(1)
    return opts["url"], strategy


def main():
    url, strategy = parse_args(sys.argv[1:])

    if not url:
        print("Usage: PAGESPEED_API_KEY=<key> python pagespeed-audit.py [--url <url>] [--strategy mobile|desktop]")
        print("Example: PAGESPEED_API_KEY=AIzaSy... python pagespeed-audit.py --url https://example.com --strategy desktop")
        sys.exit(1)

    api_key = os.environ.get("PAGESPEED_API_KEY")
    if not api_key:
        print("Error: PAGESPEED_API_KEY environment variable is missing.", file=sys.stderr)
        print("Agent Action Required: Please ask the user to provide their Google PageSpeed API key, or read it from a .env file, then rerun the command with the key.", file=sys.stderr)
        sys.exit(1)

    api_url = (
        f"https://www.googleapis.com/pagespeedonline/v5/runPagespeed"
        f"?url={urllib.parse.quote(url)}"
        f"&strategy={strategy.upper()}"
        f"&category=performance&category=seo&category=accessibility&category=best-practices"
        f"&key={api_key}"
    )

    print(f"Auditing {url} ({strategy} strategy)...")

    try:
        with urllib.request.urlopen(api_url) as response:
            data = response.read()
            result = json.loads(data)

            if "error" in result:
                print(f"Error from API: {result['error']['message']}")
                sys.exit(1)

            categories = result["lighthouseResult"]["categories"]
            audits = result["lighthouseResult"]["audits"]

            def cat_score(cat_id):
                cat = categories.get(cat_id)
                if cat and cat.get("score") is not None:
                    return round(cat["score"] * 100)
                return None

            perf_score = cat_score("performance")
            seo_score = cat_score("seo")
            a11y_score = cat_score("accessibility")
            bp_score = cat_score("best-practices")

            print("\n--- PageSpeed Insights Report (Python) ---")
            print(f"URL: {url}")
            print(f"Strategy: {strategy}")
            print(f"Performance Score: {perf_score}  (target: 90+)")
            print(f"SEO Score: {seo_score}  (target: 90+)")
            if a11y_score is not None:
                print(f"Accessibility Score: {a11y_score}  (target: 90+)")
            if bp_score is not None:
                print(f"Best Practices Score: {bp_score}  (target: 90+)")

            # CrUX field data (real-user Core Web Vitals, 28-day window)
            le_metrics = (result.get("loadingExperience") or {}).get("metrics") or {}
            if le_metrics:
                overall = result["loadingExperience"].get("overall_category", "n/a")
                print(f"\nCore Web Vitals (Field Data, {overall}):")
                for mid, m in le_metrics.items():
                    print(f"- {mid}: {m['percentile']} ({m['category']})")
            else:
                print("\nCore Web Vitals (Field Data): no CrUX data (traffic below threshold) — relying on lab data below")

            print("\nKey Metrics (Lab):")

            def metrics_val(aid):
                return (audits.get(aid) or {}).get("displayValue") or "n/a"

            print(f"- First Contentful Paint: {metrics_val('first-contentful-paint')}")
            print(f"- Largest Contentful Paint: {metrics_val('largest-contentful-paint')}")
            print(f"- Cumulative Layout Shift: {metrics_val('cumulative-layout-shift')}")
            print(f"- Total Blocking Time: {metrics_val('total-blocking-time')}")
            print(f"- Speed Index: {metrics_val('speed-index')}")

            # Failing audits (score < 0.9) across all categories
            EXCLUDE = {
                "metrics", "screenshot-thumbnails", "final-screenshot", "timing",
                "network-requests", "network-rtt", "network-server-latency",
                "diagnostics", "debug-data", "stacks", "errors-in-console",
                "worker-timing", "lcp-lazy-loaded", "prioritize-lcp-image",
                "uses-rel-preconnect", "third-party-summary",
                "first-contentful-paint", "largest-contentful-paint",
                "cumulative-layout-shift", "total-blocking-time", "speed-index",
                "interactive",
            }
            weighted_ids = {
                ref["id"]
                for cat in categories.values()
                for ref in cat.get("auditRefs", [])
                if ref.get("weight", 0) > 0
            }
            failing = []
            for aid, audit in audits.items():
                score = audit.get("score")
                if score is None or score >= 0.9 or aid in EXCLUDE:
                    continue
                failing.append({
                    "id": aid,
                    "score": score,
                    "displayValue": audit.get("displayValue", ""),
                    "weighted": aid in weighted_ids,
                })
            failing.sort(key=lambda f: f["score"])
            if failing:
                print("\nFailing Audits (score < 0.9):")
                for f in failing[:15]:
                    tag = " [core-weighted]" if f["weighted"] else ""
                    print(f"- {f['id']}: {f['displayValue'] or 'failed'}{tag}")
                    fix = FIXES.get(f["id"])
                    if fix:
                        print(f"  -> Fix: {fix}")

            # Savings opportunities (bytes/ms estimates)
            SAVINGS = [
                "unused-javascript", "unused-css-rules", "render-blocking-resources",
                "modern-image-formats", "uses-optimized-images", "offscreen-images",
                "uses-responsive-images", "uses-long-cache-ttl", "legacy-javascript",
                "duplicated-javascript", "font-display",
            ]
            opportunities = [
                audits[sid] for sid in SAVINGS
                if sid in audits and audits[sid].get("score") is not None and audits[sid]["score"] < 1
            ]
            opportunities.sort(key=lambda a: a["score"])
            if opportunities:
                print("\nSavings Opportunities:")
                for o in opportunities:
                    print(f"- {o['id']}: {o.get('displayValue') or 'see report'}")
                    fix = FIXES.get(o["id"])
                    if fix:
                        print(f"  -> Fix: {fix}")
            print("------------------------------------------\n")

            scores = [s for s in (perf_score, seo_score, a11y_score, bp_score) if s is not None]
            if any(s < 90 for s in scores):
                weak = sum(1 for s in scores if s < 90)
                plural = "y is" if weak == 1 else "ies are"
                print(f"Action Item: {weak} categor{plural} below the 90 target. Advise the Experience Architect to prioritize the failing audits above.")
            else:
                print("Status: All category scores meet the 90+ target. Visibility targets met.")

    except urllib.error.URLError as e:
        print(f"Request failed: {e.reason}")
        sys.exit(1)
    except Exception as e:
        print(f"Audit failed: {e}")
        sys.exit(1)

if __name__ == "__main__":
    main()
